zc_common/remote_resource/request.py

```diff
@@ -80,19 +80,30 @@ def add_items_from_data(self):
         map(lambda x: self.append(RemoteResourceWrapper(x)), self.data)
 
 
-def get_route_from_fk(resource_type, pk=None):
-    """Gets a fully qualified URL for a given resource_type, pk"""
-    routes = requests.get(zc_settings.GATEWAY_ROOT_PATH).json()
-
-    for route in routes.iterkeys():
-        if 'resource_type' in routes[route] and routes[route]['resource_type'] == resource_type:
-            if isinstance(pk, (list, set)):
-                expanded = '{}?filter[id__in]={}'.format(expand(route, {}), ','.join([str(x) for x in pk]))
-            else:
-                expanded = expand(route, {'id': pk})
-            return '{0}{1}'.format(routes[route]['domain'], expanded)
-
-    raise RouteNotFoundException('No route for resource_type: "{0}"'.format(resource_type))
+class RouteManager(object):
```

Comment:
Before this change, any time you called `get_route_from_fk`, you would make a request to the Gateway. This meant that if your API response has a multiple remote FKs (OrderItem has 2), you would make multiple redundant requests for each field in the response. Even just 1 request is unacceptable within the request scope, because it harshly affects response times. That means the only acceptable solution would be to cache the route list outside of the request cycle. `RouteManager` is the simplest solution to the problem without building a complex solution involving memcache or redis and a worker process to refresh the route list on a timer. Our route lists are not currently changing all that often, and services are not dynamically reacting to new routes, so this is not a problem we need to solve.

Now to address your points:

1. Anytime a service is modified to access a new route, it's being restarted anyways as part of the deployment process. This doesn't change anything. You just have to make sure the gateway is updated first.
2. Even if services register their routes automatically, that's only one part of the problem. The other part is somehow allowing services to stay up to date with the route list. A solution to this problem is still a long ways away, and there is no proposed solution for the 2nd half of the problem yet. In the meantime, `RouteManager` is an acceptable stopgap.
3. You're not modifying the gateway's route list in the middle of tests, so this changes nothing. In fact, this will speed up tests over the previous version considerably for the reasons mentioned above.
4. Caching at request time is unacceptable because it affect's that request's response time. Again, like I said in the previous point, unit tests are unaffected. Refreshing on a timer is impossible without a worker process and redis/memcached, and as I said above, that's way more complicated than we need right now.